SCClassLibrary/Common/Control/ServerNodeTree.sc

```diff
@@ -0,0 +1,198 @@
+ServerNodeTree {
+
+	var server;
+	var updateFunc, updater, <>refreshRate = 0.2;
+	var currentMsg, currentNodeTree = nil;
+	var <running = false;
+	var oscResp;
+	
+	*new { |server|
+		^super.new.init(server)
+	}
+	
+	init { |parentServer|
+		server = parentServer;
+	}
+
+	getNodeTree {
+		if(running) {
+			if(currentNodeTree.isNil) {
+				this.prFormatNodeTree;
+			};
+			^currentNodeTree
+		} {
+			^nil
+		};
+	}
+
+	run {
+		if(running.not) {
+			// Init osc receiver
+			oscResp = OSCFunc(
+				{ |msg|
+					// Only catch relevant messages
+					// (root node and controls enabled)
+					if((msg[1] == 1) and: { msg[2] == 0 }) {
+						this.prClearNodeTree;
+						currentMsg = msg;
+					};
+				},
+				'/g_queryTree.reply',
+				server.addr
+			).fix;
+			// Start auto update thread
+			updateFunc = {
+				updater = fork {
+					loop {
+						server.sendMsg(
+							"/g_queryTree",
+							0,
+							1
+						);
+						refreshRate.wait;
+					};
+				};
+				running = true;
+			};
+		};
+		updateFunc.value;
+		CmdPeriod.remove(updateFunc);
+	}
+
+	stop {
+		if(running) {
+			oscResp.free;
+			updater.stop;
+			CmdPeriod.remove(updateFunc);
+			running = false;
+		};
+	}
+
+	/*    PRIVATE METHODS    */
+
```

Comment:
I'm not well-versed in garbage collection art. I think it can indeed be garbage collected, but over multiple passes. I think that when it first passes, it marks top level as de-referenced, but sub-levels are still referenced in top-level, so they won't be marked until next garbage collection. So one pass for every level. Clearing the whole tree allows to remove everything in a single pass?